test/contract.js

```diff
@@ -9,231 +9,247 @@ var StandAloneContract = require('../packages/web3-eth-contract');
-    'inputs': [],
-    'constant': true,
-    'outputs': [
-        {'name': '', 'type': 'uint256'}
+    "name":"overloadedFunction",
+    "type":"function",
+    "inputs":[],
+    "constant":true,
+    "outputs":[
+        {"name":"","type":"uint256"}
     ],
-    'payable': false,
-    'stateMutability': 'view'
-}];
+    "payable":false,
+    "stateMutability":"view"
+    }];
 
 var address = '0x11f4d0A3c12e86B4b5F39B213F7E19D048276DAe';
 var addressLowercase = '0x11f4d0a3c12e86b4b5f39b213f7e19d048276dae';
 var address2 = '0x5555567890123456789012345678901234567891';
-var account = {
-    address: '0xEB014f8c8B418Db6b45774c326A0E64C78914dC0',
-    privateKey: '0xbe6383dad004f233317e46ddb46ad31b16064d14447a95cc1d8c8d4bc61c3728'
-};
 
-/**
- * Factory method for standalone Contract objects.
- *
- * @method getStandAloneContractInstance
- *
- * @param {Array} abi
- * @param {String} address
- * @param {providers} provider
- * @param {Object} options
- *
- * @returns {Contract}
- */
-var getStandAloneContractInstance = function (abi, address, provider, options) {
+var getStandAloneContractInstance = function(abi, address, options, provider) {
+
+    // if no address supplied
+    if (address && typeof address != 'string') {
+
+        // no options provided, either
+        if (!options && !provider) {
+            provider = address;
+            address = undefined;
+            options = undefined;
+        } else if (options) {
+            // options provided, but no address
+            options = address;
+            provider = options;
+            address = undefined;
+        }
+    } else if (!provider) {
+        // address provided, but no options
+        provider = options;
+        options = undefined;
+    }
+
+
     StandAloneContract.setProvider(provider);
     return new StandAloneContract(abi, address, options);
-};
+}
+
+var getEthContractInstance = function(abi, address, options, provider) {
+
+    // if no address supplied
+    if (address && typeof address != 'string') {
+
+        // no options provided, either
+        if (!options && !provider) {
+            provider = address;
+            address = undefined;
+            options = undefined;
+        } else if (options) {
+            // options provided, but no address
+            options = address;
+            provider = options;
+            address = undefined;
+        }
+    } else if (!provider) {
+        // address provided, but no options
+        provider = options;
+        options = undefined;
+    }
 
-/**
- * Factory method for eth.Contract objects.
- *
- * @method getEthContractInstance
- *
- * @param {Array} abi
- * @param {String} address
- * @param {providers} provider
- * @param {Object} options
- *
- * @returns {Contract}
- */
-var getEthContractInstance = function (abi, address, provider, options) {
     var eth = new Eth(provider);
-    const contract = new eth.Contract(abi, address, options);
+    //eth.setProvider(provider);
```

Comment:
The call of ``setProvider`` is not required because the ``constructor`` of the ``Eth`` module does set the provider already as expected. The tests fail because the added validations to the ``FakeIpcProvider`` aren't in the correct order (I will check that closer). But we can say that the logic behind ``setProvider`` is still correct or the test case on line 3112 would fail.

Reply:
Yes, it seems possible everything is ok but it would be good to certain as well :)

Reply:
@cgewecke I've checked the issue we have with the additional call of ``eth.setProvider`` closer. Because we pass the exact same provider again with ``setProvider`` does it register the ``data`` listener again and the ``eth_getTransactionReceipt`` call will get executed more than required. We could implement the removal of all listeners on ``setProvider`` to prevent this behaviour but this would probably break existing DApps which do register event listeners before ``setProvider`` got called. 

After all, if we add an explanatory note to the ``setProvider`` method to inform the user about this behaviour can we ignore it. WDYT?

Reply:
Ok - that sounds reasonable...I'm going to walk the code again as you describe just to satisfy my paranoia.

Reply:
@cgewecke Great! 👌